# SQL Exercise (10 points)

# Part 1

Complete this exercise using SQLite. 

The exercise does not require GPU, therefore you can use your local machine. To work with SQLite, first install `ipython-sql` (e.g., 'pip install ipython-sql').

Or, you can also use Google Colab if you wish, which has the `ipython-sql` installed.

1.1 Create the following table named `Products` that has 4 columns with 10 records. Set the `Code` column to be the Primary Key. 

In [4]:
# Your code here

 * sqlite://
Done.


[(1, 'Hard drive', 240, 'Fujitsu'),
 (2, 'Memory', 120, 'Winchester'),
 (3, 'ZIP drive', 150, 'Iomega'),
 (4, 'Floppy disk', 5, 'Winchester'),
 (5, 'Monitor', 240, 'Sony'),
 (6, 'DVD drive', 180, 'Creative Labs'),
 (7, 'CD drive', 90, 'Creative Labs'),
 (8, 'Printer', 270, 'Hewlett-Packard'),
 (9, 'Toner cartridge', 66, 'Hewlett-Packard'),
 (10, 'DVD burner', 180, 'Creative Labs')]

If you have many errors, sometimes it helps to restart the kernel by clicking on `Kernel` -> `Restart Kernel` in the top menu, and run again the cells. 

1.2 Select the names of all the products in the store.

In [5]:
# Your code here

 * sqlite://
Done.


[('Hard drive',),
 ('Memory',),
 ('ZIP drive',),
 ('Floppy disk',),
 ('Monitor',),
 ('DVD drive',),
 ('CD drive',),
 ('Printer',),
 ('Toner cartridge',),
 ('DVD burner',)]

1.3 Select the names of the products with a price less than or equal to $200. Display product names and prices.

In [ ]:
# Your code here

1.4 Select all products with a price between $60 and 200. Display product names and prices, sorted by ascending price.

In [7]:
# Your code here

 * sqlite://
Done.


[('Toner cartridge', 66),
 ('CD drive', 90),
 ('Memory', 120),
 ('ZIP drive', 150),
 ('DVD drive', 180),
 ('DVD burner', 180)]

1.5 Select the name and price of all products, and display the price shown in cents (instead of dollars).

In [9]:
# Your code here

 * sqlite://
Done.


[('Hard drive', 24000),
 ('Memory', 12000),
 ('ZIP drive', 15000),
 ('Floppy disk', 500),
 ('Monitor', 24000),
 ('DVD drive', 18000),
 ('CD drive', 9000),
 ('Printer', 27000),
 ('Toner cartridge', 6600),
 ('DVD burner', 18000)]

1.6 Compute the average price of all products from the manufacturer 'Creative Labs'.

In [10]:
# Your code here

 * sqlite://
Done.


[(150.0,)]

1.7. Compute the number of products with a price larger than or equal to $180.

In [11]:
# Your code here

 * sqlite://
Done.


[(5,)]

1.8 Select the name and price of all products with a price larger than or equal to $180, and sort them by price in descending order.

In [ ]:
# Your code here

1.8 Select the manufacturers and the average price of each manufacturer's products.

In [13]:
# Your code here

 * sqlite://
Done.


[('Creative Labs', 150.0),
 ('Fujitsu', 240.0),
 ('Hewlett-Packard', 168.0),
 ('Iomega', 150.0),
 ('Sony', 240.0),
 ('Winchester', 62.5)]

1.9 Select the manufacturers whose products have an average price larger than or equal to $160.

In [14]:
# Your code here

 * sqlite://
Done.


[(240.0, 'Fujitsu'), (168.0, 'Hewlett-Packard'), (240.0, 'Sony')]

1.10 Select the name and price of the cheapest product.

In [ ]:
# Your code here

1.11 Select the name of each manufacturer along with the name and price of its most expensive product, and sort them by price in descending order.

In [16]:
# Your code here

 * sqlite://
Done.


[('Hewlett-Packard', 270),
 ('Sony', 240),
 ('Fujitsu', 240),
 ('Creative Labs', 180),
 ('Iomega', 150),
 ('Winchester', 120)]

1.12 Update the name of product 8 to "Laser Printer", and display the updated names. 

In [ ]:
# Your code here

1.13 Apply a 10% discount to all products. Display the product names and updated prices. 

In [ ]:
# Your code here

1.14 Apply an additional 20% discount to all products with a price larger than or equal to $120.

In [ ]:
# Your code here

Delete the table from Part 1.

In [ ]:
# Your code here

# Part 2

Run the code in the following cell to create a database containing information about employees and departments in a small business, and afterward answer the following questions. 

In [24]:
%%sql sqlite://
CREATE TABLE Departments (
  Code INTEGER PRIMARY KEY,
  Name varchar(255) NOT NULL ,
  Budget decimal NOT NULL 
);

CREATE TABLE Employees (
  SSN INTEGER PRIMARY KEY,
  Name varchar(255) NOT NULL ,
  LastName varchar(255) NOT NULL ,
  Department INTEGER NOT NULL , 
  foreign key (department) references Departments(Code) 
);

INSERT INTO Departments(Code,Name,Budget) VALUES(14,'IT',65000);
INSERT INTO Departments(Code,Name,Budget) VALUES(37,'Accounting',15000);
INSERT INTO Departments(Code,Name,Budget) VALUES(59,'Human Resources',240000);
INSERT INTO Departments(Code,Name,Budget) VALUES(77,'Research',55000);

INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('123234877','Michael','Rogers',14);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('152934485','Anand','Manikutty',14);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('222364883','Carol','Smith',37);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('326587417','Joe','Stevens',37);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('332154719','Mary-Anne','Foster',14);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('332569843','George','ODonnell',77);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('546523478','John','Doe',59);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('631231482','David','Smith',77);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('654873219','Zacary','Efron',59);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('745685214','Eric','Goldsmith',59);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('845657245','Elizabeth','Doe',14);
INSERT INTO Employees(SSN,Name,LastName,Department) VALUES('845657246','Kumar','Swamy',14);

Done.
Done.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.
1 rows affected.


[]

2.1 Select the last name of all employees.

In [25]:
# Your code here

 * sqlite://
Done.


[('Rogers',),
 ('Manikutty',),
 ('Smith',),
 ('Stevens',),
 ('Foster',),
 ('ODonnell',),
 ('Doe',),
 ('Smith',),
 ('Efron',),
 ('Goldsmith',),
 ('Doe',),
 ('Swamy',)]

2.2 Select the last name of all employees, without duplicates. Sort them in alphabetical order. 

In [ ]:
# Your code here

2.3 Select all the data of employees whose last name is "Smith" or "Doe".

In [27]:
# Your code here

 * sqlite://
Done.


[(222364883, 'Carol', 'Smith', 37),
 (546523478, 'John', 'Doe', 59),
 (631231482, 'David', 'Smith', 77),
 (845657245, 'Elizabeth', 'Doe', 14)]

2.4 Select all the data of employees whose last name begins with an "S".

In [28]:
# Your code here

 * sqlite://
Done.


[(222364883, 'Carol', 'Smith', 37),
 (326587417, 'Joe', 'Stevens', 37),
 (631231482, 'David', 'Smith', 77),
 (845657246, 'Kumar', 'Swamy', 14)]

2.5 Select the number of employees in each department, and display the name of each department and the number of employees per department. 

In [ ]:
# Your code here

2.6 Select the name and last name of employees working for the departments that have a budget greater than $60,000.

In [30]:
# Your code here

 * sqlite://
Done.


[('Michael', 'Rogers'),
 ('Anand', 'Manikutty'),
 ('Mary-Anne', 'Foster'),
 ('John', 'Doe'),
 ('Zacary', 'Efron'),
 ('Eric', 'Goldsmith'),
 ('Elizabeth', 'Doe'),
 ('Kumar', 'Swamy')]

2.7 Select the names of departments with more than two employees (e.g., you can use a subquery).

In [ ]:
# Your code here

2.8 Reduce the budget of all departments by 10%. Display the departments table to verify the changes. 

In [ ]:
# Your code here

2.9 Reassign all employees from the Research department (code 77) to the IT department (code 14). Display the employees table to verify the change.

In [ ]:
# Your code here

Delete the tables in the current database. 

In [ ]:
# Your code here

# Part 3

3.1 Connect to the database named `demo.db` located in the `data` folder. 

In [38]:
# Your code here

3.2 Select the names of all tables in the database. 

In [39]:
# Your code here

   sqlite://
 * sqlite:///data/demo.db
Done.


[('Employee',),
 ('Category',),
 ('Customer',),
 ('Shipper',),
 ('Product',),
 ('OrderDetail',)]

3.3 Select the names of all categories in the Category table. 

In [40]:
# Your code here

   sqlite://
 * sqlite:///data/demo.db
Done.


[(1, 'Beverages', 'Soft drinks, coffees, teas, beers, and ales'),
 (2, 'Condiments', 'Sweet and savory sauces, relishes, spreads, and seasonings'),
 (3, 'Confections', 'Desserts, candies, and sweet breads'),
 (4, 'Dairy Products', 'Cheeses'),
 (5, 'Grains/Cereals', 'Breads, crackers, pasta, and cereal'),
 (6, 'Meat/Poultry', 'Prepared meats'),
 (7, 'Produce', 'Dried fruit and bean curd'),
 (8, 'Seafood', 'Seaweed and fish')]

3.4 Select the names of all categories, description of categories, and the number of products per category (e.g, use the `CategoryID` column in the `Product` table to relate to the `ID` column in the `Category` table).

In [41]:
# Your code here

   sqlite://
 * sqlite:///data/demo.db
Done.


[('Beverages', 'Soft drinks, coffees, teas, beers, and ales', 12),
 ('Condiments', 'Sweet and savory sauces, relishes, spreads, and seasonings', 12),
 ('Confections', 'Desserts, candies, and sweet breads', 13),
 ('Dairy Products', 'Cheeses', 10),
 ('Grains/Cereals', 'Breads, crackers, pasta, and cereal', 7),
 ('Meat/Poultry', 'Prepared meats', 6),
 ('Produce', 'Dried fruit and bean curd', 5),
 ('Seafood', 'Seaweed and fish', 12)]

3.5 From the `Customer` table, select the names of 3 countries with the largest number of customers per country (e.g., use the `Country` column in the `Customer` table, and create a subquery for the number of customers per country).

In [42]:
# Your code here

   sqlite://
 * sqlite:///data/demo.db
Done.


[('USA', 13), ('France', 11), ('Germany', 11)]

3.6 Count the number of discontinued products (in the `Product` table, 1 in the `Discontinued` column means that the product is discontinued).

In [43]:
# Your code here

   sqlite://
 * sqlite:///data/demo.db
Done.


[(8,)]

3.7 Select products that cost between $20 and 25 and are not discontinued. Display product name and unit price, sorted by price in descending order. 

In [ ]:
# Your code here

3.8 Select the five most expensive products, and display the product name and unit price. 

In [45]:
# Your code here

   sqlite://
 * sqlite:///data/demo.db
Done.


[('Côte de Blaye', 263.5),
 ('Thüringer Rostbratwurst', 123.79),
 ('Mishi Kobe Niku', 97),
 ("Sir Rodney's Marmalade", 81),
 ('Carnarvon Tigers', 62.5)]